### PR TITLE
lredir: Ask DOS if drive is free [fixes #768]

### LINKED ISSUE #768: Booting with -K and -I "disk { hdimage ... }" shadows one of the drives used for disk image partitions
**Describe the bug**
When booting with both a -K directory and -I "disk { hdimage ... }", drive G is first used for a partition in the disk image, but then seems to also be used to redirect to the -K directory.

**To Reproduce**

```
$ rm fat-128m.img
$ truncate -s "$((128 * 1024 * 1024))" fat-128m.img
$ /sbin/fdisk fat-128m.img

Welcome to fdisk (util-linux 2.33.1).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS disklabel with disk identifier 0xa4de813e.

Command (m for help): n
Partition type
   p   primary (0 primary, 0 extended, 4 free)
   e   extended (container for logical partitions)
Select (default p): 

Using default response p.
Partition number (1-4, default 1): 
First sector (2048-262143, default 2048): 
Last sector, +/-sectors or +/-size{K,M,G,T,P} (2048-262143, default 262143): 

Created a new partition 1 of type 'Linux' and of size 127 MiB.

Command (m for help): t
Selected partition 1
Hex code (type L to list all codes): 6
Changed type of partition 'Linux' to 'FAT16'.

Command (m for help): w
The partition table has been altered.
Syncing disks.

$ dosemu -dumb -K "$PWD" -I "disk { hdimage fat-128m.img }"
ERROR: hdimage size is not cylinder-aligned (5104 sectors), truncated!
                                                                      dosemu2 2.0pre8-20190125-883-g8318127b4
Configured: 2019-01-14 13:29:43 +0100
Please test against a recent version before reporting bugs and problems.
Get the latest code at http://stsp.github.io/dosemu2
Submit Bugs via https://github.com/stsp/dosemu2/issues
Ask for help in mail list: linux-msdos@vger.kernel.org

FreeDOS kernel - SVN (build 2042 OEM:0xfd) [compiled Sep 22 2017]
Kernel compatibility 7.10 - BORLANDC - FAT32 support

(C) Copyright 1995-2012 Pasquale J. Villani and The FreeDOS Project.
All Rights Reserved. This is free software and comes with ABSOLUTELY NO
WARRANTY; you can redistribute it and/or modify it under the terms of the
GNU General Public License as published by the Free Software Foundation;
either version 2, or (at your option) any later version.
C: HD1, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
D: HD2, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
E: HD3, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
F: HD4, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
G: HD5, Pri[ 1], CHS=    0-32-33, start=     1 MB, size=   127 MB
dosemu XMS 3.0 & UMB support enabled                                           
dosemu EMS driver rev 0.8 installed.
EMUFS host file and print access available
Kernel: allocated 30 Diskbuffers = 15960 Bytes in HMA
BLASTER=A220 I5 D1 H5 P330 T6
MIDI=SYNTH:2 MAP:E MODE:0
Welcome to dosemu2!
    Build 2.0pre8-20190125-883-g8318127b4
G:\>lredir
Current Drive Redirections:
C: = \\LINUX\FS\[home]\.dosemu\drive_c\ attrib = READ/WRITE
D: = \\LINUX\FS\usr\local\share\dosemu\dosemu2-cmds-0.2\ attrib = READ/WRITE
E: = \\LINUX\FS\usr\local\share\fdpp\ attrib = READ/WRITE
F: = \\LINUX\FS\usr\local\share\dosemu\freedos\ attrib = READ/WRITE
G: = \\LINUX\FS\[home]\test\20190131\ attrib = READ/WRITE
G:\>dir
 Volume in drive G is ST 20190131
 Directory of G:\

HARD     IMG    536,870,912  01-31-19  1:18p
FLOPP~V4 CLE     1,474,560  01-31-19  1:24p
FORMAT   ZIP        31,218  01-31-19 12:57p
HARDI~JN CLE    536,870,912  01-31-19  1:18p
FLOPPY   IMG     1,474,560  01-31-19  1:24p
FAT-128M IMG    134,217,728  01-31-19  1:29p
         6 file(s)   1,210,939,890 bytes
         0 dir(s)      2,588 Mega bytes free
G:\>exitemu
$ 
```

**Attach the log**
[boot.log](https://github.com/stsp/dosemu2/files/2817156/boot.log)

### PR DESCRIPTION
**Pull request by andrewbird**
### Updated commit message

Change check to test for existence those physical drives that are
present but unformatted.

 ##### The test below has:
  - one fatfs virtual disk
  - two image disks with multiple primary partitions
  - one image disk with one unformatted primary partition, an
    extended partition, and an unformatted logical partition.
  - three automatic fatfs virtual disks

$_hdimage = "dXXXXs/c:hdtype1 dXXXXs/hard1.img dXXXXs/hard2.img"
            "dXXXXs/unformatted.img +1"

C: HD1, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=    10 MB
D: HD2, Pri[ 1], CHS=    0-32-33, start=     1 MB, size=    12 MB
E: HD3, Pri[ 1], CHS=    0-32-33, start=     1 MB, size=    12 MB
F: HD4, Pri[ 1], CHS=    0-32-33, start=     1 MB, size=     5 MB
G: HD5, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
H: HD6, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
I: HD7, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
J: HD4, Ext[ 1], CHS=    0-227-36, start=     7 MB, size=     3 MB
K: HD2, Pri[ 2], CHS=    1-167-39, start=    13 MB, size=   128 MB
L: HD2, Pri[ 3], CHS=   17-248-40, start=   141 MB, size=   371 MB
M: HD3, Pri[ 2], CHS=    1-167-39, start=    13 MB, size=   128 MB
N: HD3, Pri[ 3], CHS=   17-248-40, start=   141 MB, size=   371 MB

 ##### Avoids all the drives above and allocates the next available
C:\>lredir -n /tmp
O: = \\LINUX\FS/TMP attrib = READ/WRITE
C:\>

### CODE REVIEW

**Comment by stsp** [comment]
And what about logical unformatted
partitions?

**Comment by andrewbird** [comment]
I repartitioned an image as follows
~~~
test-imagedir/dXXXXs/unformatted.img1       2048 12287   10240    5M  1 FAT12
test-imagedir/dXXXXs/unformatted.img2      12288 20807    8520  4.2M  5 Extended
test-imagedir/dXXXXs/unformatted.img5      14336 20807    6472  3.2M  1 FAT12
~~~

then did
~~~
C: HD1, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=    10 MB
D: HD2, Pri[ 1], CHS=    0-32-33, start=     1 MB, size=    12 MB
E: HD3, Pri[ 1], CHS=    0-32-33, start=     1 MB, size=    12 MB
WARNING: using suspect partition Pri:1 FS 01: with calculated values    0-32-33 instead of    1-37-15
WARNING: using suspect partition Pri:1 FS 01: with calculated values    0-195-3 instead of    8-74-12
F: HD4, Pri[ 1], CHS=    0-32-33, start=     1 MB, size=     5 MB
G: HD5, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
H: HD6, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
I: HD7, Pri[ 1], CHS=    0-1-1, start=     0 MB, size=  2000 MB
WARNING: using suspect partition Ext:1 FS 01: with calculated values    0-227-36 instead of   10-36-9
WARNING: using suspect partition Ext:1 FS 01: with calculated values    1-75-18 instead of   15-15-18
J: HD4, Ext[ 1], CHS=    0-227-36, start=     7 MB, size=     3 MB
K: HD2, Pri[ 2], CHS=    1-167-39, start=    13 MB, size=   128 MB
L: HD2, Pri[ 3], CHS=   17-248-40, start=   141 MB, size=   371 MB
M: HD3, Pri[ 2], CHS=    1-167-39, start=    13 MB, size=   128 MB
N: HD3, Pri[ 3], CHS=   17-248-40, start=   141 MB, size=   371 MB

C:\>lredir -n /tmp
O: = \\LINUX\FS/TMP attrib = READ/WRITE

C:\>lredir -n /bin
P: = \\LINUX\FS/BIN attrib = READ/WRITE
C:\>
~~~

So that looks okay to me.

**Comment by stsp** [comment]
So why do we need 2 checks instead
of one then?
Have you tried serial number as a
single check?

**Comment by stsp** [comment]
Ah you already wrote an explanation
elsewhere...

**Comment by andrewbird** [comment]
No I didn't try because you can't get serial number from a network drive so you'll need to check twice for that, and for the physical drives it reads it from the BPB which is initialised during format.

**Comment by stsp** [comment]
Ok I suppose if the second check is
only needed for network drives, then
it can be done explicitly rather than
via free space query?

**Comment by andrewbird** [comment]
Let me see what function's available for that.

**Comment by andrewbird** [comment]
Only way I can see that should be safe with multiple redirectors would be to iterate through com_GetRedirection() results checking for drive letter, which seems like a worse solution.

On a slightly related note I think we need to figure out GetRedirection() index behaviour with multiple redirectors/clients definitively.

**Comment by stsp** [comment]
Redirector can deal with drive letters
directly, so involving index looks wrong.

**Comment by andrewbird** [comment]
> Redirector can deal with drive letters
directly, so involving index looks wrong.

So which function allows me to ask if a drive is mapped as going around the back and accessing the drives array is not right?

**Comment by stsp** [comment]
I am not at a pc right now, but what I
do remember is that select_drive()
precedes most of the redirector funtions,
and returns error if not found.
Qualify_filename?
Also perhaps you can just do change
drive to see if it is there?

**Comment by stsp** [comment]
Would be much better if we could ask
dos to redirect  to free drive. If there is
no such thing, maybe at least for fdpp...

**Comment by andrewbird** [comment]
I pushed this, which uses the same trick that users use in batchfile. There's no diskinfo check, but opening and closing the nul device seems a little less robust.

**Comment by stsp** [comment]
Why not change drive instead of
both checks?

**Comment by andrewbird** [comment]
I didn't try it but it looks like [int21/0e](http://www.ctyme.com/intr/rb-2570.htm) setting the default drive always succeeds, it suggests using [int21/29](http://www.ctyme.com/intr/rb-2685.htm). But I don't think that solves the unformatted partition problem, so I expect there will still need to be two checks, although not needing to close a file might be good.

**Comment by stsp** [comment]
Its not that simple.
Set drive does not always success,
it just doesn't return error immediately.
You can use get drive to see if it was
successful. Unformatted partition should
be fine too, you just get a critical error later
when try some operation on that partition.
But if the doc suggests 29 instead -
that should be fine too.

**Comment by andrewbird** [comment]
> Unformatted partition should
be fine too, you just get a critical error later
when try some operation on that partition.

But how do you distinguish between a drive that exists and is unformatted and a drive that doesn't exist, as it seems both should fail some filesystem operation?

**Comment by stsp** [comment]
Easily: unformatted drive fails later
operation, but missing drive fails also
set drive (but you need to do get
drive to find the error)

**Comment by andrewbird** [comment]
You sound like you've coded this up already, feel free to forget this PR.

**Comment by stsp** [comment]
Yes, in comcom32 recently.
Why not to explore also 29h?

**Comment by andrewbird** [comment]
No reason, just running out of time.

**Comment by andrewbird** [comment]
Updated to use 'Parse filename in FCB' DOS v1 function, handles formatted, unformatted, network, and logical partitions in a single call.

File: src/dosext/builtins/builtins.c [hunk 1]
```diff
@@ -320,17 +320,32 @@ int com_dossetcurrentdir(char *path)
...
+  char *fname, *fcb;
+  int ret;
 
   pre_msdos();
 
-  HI(ax) = 0x36; // get free disk space
-  LO(dx) = drv;
+  /* Parse filename into FCB (physical, formatted or not, and network) */
+  fname = lowmem_heap_alloc(16);
+  snprintf(fname, 16, "%c:FILENAME.EXT", 'A' - 1 + drv);
```
Comment (stsp) [inline src/dosext/builtins/builtins.c]: Maybe use "NUL" here?

Comment (andrewbird) [reply src/dosext/builtins/builtins.c]: Could do but it's not testing for existence, just that the name is parseable and the drive is valid.

**Comment by stsp** [comment]
Thanks!

File: src/dosext/builtins/builtins.c (hunk 1 above)
Comment (stsp) [reply src/dosext/builtins/builtins.c]: Is there even no need for `\` after `:`?

Comment (andrewbird) [reply src/dosext/builtins/builtins.c]: Actually it won't work if you do that, it's a DOS v1 function and that had no directories

**Pull request merged**